Anni/wcicd (#16642)
cloudflare/cloudflare-docs: src/content/docs/workers/ci-cd/index.mdx

---
pcx_content_type: concept
title: CI/CD
description: Set up continuous integration and continuous deployment for your Workers.
sidebar:
  order: 7

> ---

Reviewer: Something on this page should explain when to use Workers Builds, when to use Github Actions, no? Or give some guidance?

You can set up continuous integration and continuous deployment (CI/CD) for your Workers project by using either the native build system, [Workers Builds](#workers-builds), or using [third-party services](#external-cicd) to optimize your development workflow.

## Why use CI/CD?

> It is best practice to deploy your Workers with a CI/CD pipeline to automate the build and deployment process instead of running adhoc deployments manually using `wrangler deploy`. This approach ensures that your Worker is consistently built and deployed in the same environment to reduce variability. Additionally, it simplifies managing access to production credentials.

Reviewer: This section would really benefit from _short_ bullet points and to get tightened up.

## Workers Builds

[Workers Builds](/workers/ci-cd/builds) is Cloudflare's native CI/CD system that allows you to integrate with GitHub or GitLab, and automatically deploy your changes with each new commit. 

> ![Workers Builds Workflow Diagram](~/assets/images/workers/platform/ci-cd/workers-builds-workflow.png)

Reviewer: Would really benefit from a clear getting started CTA below the diagram

## External CI/CD

You can also choose to set up your CI/CD pipeline with an external provider. 

> - [GitHub Actions](https://github.com/marketplace/actions/deploy-to-cloudflare-workers-with-wrangler)

Reviewer: Why just link externally to here, if we also have a section in the docs?

https://anni-wcicd.cloudflare-docs-7ou.pages.dev/workers/ci-cd/external-cicd/

It is odd to me to be looking at this page, see a tab on the left for External CI/CD, but then click on this and get directed to somewhere external:

<img width="578" alt="Screenshot 2024-09-22 at 2 17 21 PM" src="https://github.com/user-attachments/assets/b1c7ee93-d0e2-4af3-9880-0dc45bf34c30">

Continue to think that Github and Gitlab _each_ are deserving of their own section in our docs, if for nothing else but for SEO and clarity. You can make sections that link out to somewhere external in the nav, ex:

<img width="283" alt="Screenshot 2024-09-22 at 2 19 15 PM" src="https://github.com/user-attachments/assets/63cafdc9-a237-4f12-a4f3-62e5a64395fe">

Author: made a new section for it with its own pages, i think it will help if we want to add more guides for other workflows in the future

Author: <img width="288" alt="Screenshot 2024-09-23 at 5 09 09 PM" src="https://github.com/user-attachments/assets/7ca39f89-23ce-4fc0-afd1-011ff4382b62">

---
pcx_content_type: concept
title: Builds (CI/CD)
description: Set up continuous integration and continuous deployment for your Workers.
sidebar:
  order: 7
---

> Set up **continuous integration and continuous deployment (CI/CD)** for your Workers project by using either [Workers Builds](#workers-builds) or using [existing services and platforms](#external-cicd) to optimize your development workflow.

Reviewer: Think we can be clearer and more specific than "existing services and platforms" — Github Actions, etc.

Reviewer: ```suggestion
You can set up **continuous integration and continuous deployment (CI/CD)** for your Workers project by using either [Workers Builds](#workers-builds) or using [existing services and platforms](#external-cicd) to optimize your development workflow.
```

## Why use CI/CD?

> It is good practice to only deploy your Workers from within a CI/CD environment, rather than running adhoc deployments manually using `wrangler deploy`. Among other benefits, deploying Workers from within a CI/CD environment makes sure that your Worker is built and deployed within a consistent environment, and it also makes it easier to control access to production credentials.

Reviewer 2: ```suggestion
It is good practice to only deploy your Workers from within a CI/CD environment, rather than running adhoc deployments manually using `wrangler deploy`. Deploying Workers from within a CI/CD environment offers several benefits. It ensures that your Worker is built and deployed in a consistent environment. It also simplifies the process of controlling access to production credentials.
```
Just changing it to shorter, simpler sentences

> It is good practice to only deploy your Workers from within a CI/CD environment, rather than running adhoc deployments manually using `wrangler deploy`. Among other benefits, deploying Workers from within a CI/CD environment makes sure that your Worker is built and deployed within a consistent environment, and it also makes it easier to control access to production credentials.

Reviewer: > It is good practice to only deploy your Workers from within a CI/CD environment, rather than running adhoc deployments manually using `wrangler deploy`.

This is tautological like saying "why use <thing>? because <thing> is good"

Reviewer: This misses the most important reason to do this — that every time you merge a PR, your code is automatically built and deployed.

> ## Workers Builds

Reviewer: Rather than a section for each of these, where each section is really just a link to a different page — we need to say the basics:

- Get started (built-in CI/CD, here is how you get going)
- Then, any _guidance_ on when to use which

> [Workers Builds](/workers/ci-cd/builds) allows you to integrate with your Git provider, and automatically deploy your changes when you make a push.

Reviewer: Can we please find a way to say Github / Gitlab more directly?

> ## External CI/CD

Reviewer: content below should live on the external CI/CD page.

---
pcx_content_type: concept

> title: CI/CD

Reviewer: Is this what you want to call this across the board? In the dash it is "Builds"

Can we avoid using acronyms in the sidebar?

Is there a different grouping here that makes sense? It needs own distinct top-level section?

Author: So this section includes information about both Workers build system & external ci/cd. My hesitation with 'Builds' as the high-level name is that GitLabs Pipeline and GitHub Actions are more CI/CD services not just builds. But I do agree that instead of 'Build Systems' for the Workers build system docs, it should be just 'Builds' to make it consistent with the dash.

I went with CI/CD instead of 'Continuous Integration / Continuous Delivery' as the header since it's too long, and CI/CD is a common abbreviation. Always open to suggestions tho!

Author: [ ] Change section name to 'Builds' instead of 'Build System'

Reviewer: yeah i see how it's tricky, lets talk about it

Reviewer 3: My 2 cents -- CI/CD should fall under the configuration tab rather than it's own top level nav item. I know the Configuration nav needs some re-work but it should have details about all things related to deploying that are outside of actual runtime capabilities. CI/CD is a means to build & deploy.

head: []

> description: Set up continuous integration and continuous deployment for your Workers to optimize your development workflow.

Reviewer: Fewer words, more literal, the "optimize your development workflow" doesn't add something concrete

sidebar:
  order: 7
---

Set up **continuous integration and continuous deployment (CI/CD)** for your Workers project by using either Workers build system, or using existing services and platforms to optimize your development workflow.

## Why use CI/CD?

> It is good practice to only deploy your Workers from within a CI/CD environment, rather than running adhoc deployments manually using `wrangler deploy`. Among other benefits, deploying Workers from within a CI/CD environment makes sure that your Worker is built and deployed within a consistent environment, and it also makes it easier to control access to production credentials.

Reviewer: Think we can show the _workflow_ with a diagram, show not tell

## Workers Build System

> Workers Build System allows you to integrate with your Git provider, and automatically deploy your changes when you make a push. To set up your build system, reference [Build System](/workers/ci-cd/build-system/).

Reviewer: Think another show rather than tell thing

---
pcx_content_type: concept
title: CI/CD
description: Set up continuous integration and continuous deployment for your Workers.
sidebar:
  order: 7
---

You can set up continuous integration and continuous deployment (CI/CD) for your Workers by using either the native build system, [Workers Builds](#workers-builds), or using [external providers](#external-cicd) to optimize your development workflow.

## Why use CI/CD?

It is best practice to deploy your Workers with a CI/CD pipeline to automate the build and deployment process. A CI/CD pipeline offers several advantages:

- Streamlines the build and deployment process, eliminating the need for manual adhoc `wrangler deploy` commands
- Ensures your Worker is always built and deployed in a uniform environment, reducing variability and potential errors
- Simplifies managing access to production credentials

## Which CI/CD should I use?

Choose [Workers Builds](/workers/ci-cd/builds) if you want a fully integrated solution within Cloudflare's ecosystem that requires minimal setup and configuration for GitHub or GitLab users.

> We recommend using [external CI/CD providers](/workers/ci-cd/external-cicd):

Reviewer 2: ```suggestion
We recommend using [external CI/CD providers](/workers/ci-cd/external-cicd) if:
```

> - If you have a self-hosted instance of GitHub or GitLabs, which is currently not supported in Workers Builds' git integration

Reviewer 2: ```suggestion
- You have a self-hosted instance of GitHub or GitLabs, which is currently not supported in Workers Builds' git integration
```

> - If you are using a git provider that is not GitHub or GitLab

Reviewer 2: ```suggestion
- You are using a git provider that is not GitHub or GitLab
```
